### DISCUSSION TITLE
Changing the theme through CTkOptionMenu

### DISCUSSION THREAD

**Discussion by humza-uddin** [TomSchimansky/CustomTkinter#1875] [Q&A]
Hello

`def set_default_color_theme(values: str):
    """ set color theme or load custom theme file by passing the path """
    ThemeManager.load_theme(values)
    
set_default_color_theme_optionmenu = customtkinter.CTkOptionMenu(window, width=20, height=50, corner_radius=8, values=["blue", "dark-blue", "green"],
                                                                       command=set_default_color_theme)
set_default_color_theme_optionmenu.grid(row=1, column=0, padx=1, pady=2)`

Is this right way to do it?
I previously had done this
`
def change_set_default_color_theme(values: str):
    customtkinter.set_default_color_theme(values)
    set_default_color_theme_optionmenu = customtkinter.CTkOptionMenu(window, width=20, height=50, corner_radius=8, values=["blue", "dark-blue", "green"],
                                                                       command=set_default_color_theme)
set_default_color_theme_optionmenu.grid(row=1, column=0, padx=1, pady=2)`
`

**Comment by Akascape** [comment]
You cannot change the theme directly, instead make it change after a restart.

**Reply by avalon60** [reply]
You would need to store the preference for the app to have it then read on startup (as @Akascape pointed out, you can't switch themes on the hoof). For example you could maintain preferences using one of a number of options:

- JSON preferences file;
- Use [Configparser ](https://www.tutorialspoint.com/configuration-file-parser-in-python-configparser)to maintain an ini file;
- Use a [SQLite ](https://www.geeksforgeeks.org/introduction-to-sqlite/) database.

The last one has perhaps a slightly bigger learning curve, but it is compact and provides a well defined way of storing and manipulating data. For example, by comparison, using JSON is susceptible to typos in respect of data keys, and as a result introducing bugs. Although, this may be less of a risk with smaller projects. 

I have some library functions for maintaining preferences in a SQLite database, if you are interested.

**Reply by humza-uddin** [reply]
I did remove it from the code because I could not find a possible solution. If I want to make changes to the theme without restarting what should I do, Do you have any idea? @avalon60 I haven't learned the uses of SQLite. I think JSON would be better option but I have no idea how to maintain preferences.

________
```
```
  self.entry  = customtkinter.CTkEntry(self.frame, width=500, height=50,font=customtkinter.CTkFont(size=24))
        self.entry.grid(row=1 , column=0, columnspan = 500, padx=10,pady=20)
        self.entry.bind("<KeyPress>", self.start)

self.entry.configure(fg="blue")```
```
 In this piece of code, I have noticed that CTkEntry doesn't support configure? If it doesn't support configure, how should I do it too?
 
 _____

**Reply by avalon60** [reply]
Hi @humza-uddin.

It's not a straight forward task. You would need to teardown and rebuild the widgets, to get the new theme to take effect.

One way would be to maintain a list of all the widgets you create, and use that to cycle through and destroy the widgets, and have a dedicated method/function to recreate them. This may present its own challenges, if you don't destroy them in the correct order, although I might be wrong on this particular point. For example if you attempt to destroy a button, after you have already destroyed its parent widget. You could of course catch and ignore the raised exceptions. A possible alternative, where for example all your widgets are contained within frames, you could use something like:

```
  for widget in frame.winfo_children():
      widget.destroy()
```

Another consideration is that you can't destroy the CTk() master widget. CTk() has an fg_color property, so you would need to do a look-up (there's a CTkThemeManager api which facilitates this.) and grab the colour code from the new theme, and explicitly configure the CTk() widget, to update it to the new colour.

Finally, much depends on your application design. Depending on what your app is doing, you may need to store the session state (data changes etc. made by the user). You would then need to restore these, after rebuilding the destroyed widgets. You might be lucky, and have an app, where CTk() itself, doesn't directly host any data widgets.
